### DISCUSSION TITLE
Permit search engines to index Github wikis

### DISCUSSION THREAD

**Discussion by ishepherd** [community/community#4992] [Other Feature Feedback, Questions, & Ideas] [answered]
There is useful content in Github wikis.

It's unfortunate that external search engines are banned ~in https://github.com/robots.txt~ 
_Edit: no longer banned in robots.txt [but still banned](https://github.com/github/feedback/discussions/4992#discussioncomment-1448177)_

People have put time into building their wikis and it's sad that the information is almost impossible to discover - unless you think to try Github search.

**Comment by nelsonjchen** [comment]
For reference here's an old issue from the old isaacs repo: https://github.com/isaacs/github/issues/1683

**Reply by nelsonjchen** [reply]
Another old discussion, which pretty much is a repeat: https://github.community/t/request-github-open-up-project-wiki-pages-to-web-indexes/122096

**Reply by nelsonjchen** [reply]
FYI @EdVassie

**Comment by nelsonjchen** [comment]
It also blocks archive.org.

**Comment by asdfugil** [comment]
they are not banned, robots.txt does nothing to their ability to crawl. It is just that they should not crawl and they don't.

**Reply by rfjakob** [reply]
I'm not sure what you mean, but line 11 in https://github.com/robots.txt is literally this:

    Disallow: /*/wiki*

**Reply by timothywcrane** [reply]
Disallow: /*/wiki* is just a directive... if you build your crawler to ignore directives... there is no tech involved trying to stop your crawl provided by the robots.txt file.

**Reply by mck1117** [reply]
@timothywcrane the question is not about us writing some crawler to crawl GitHub Wikis (because you could just ignore `robots.txt` and crawl it anyway), but that we want *Google* to crawl our wikis so that our users/contributors can find the wiki.  Google respects `robots.txt` (shocker, I know), so the only way to get a GitHub Wiki indexed by Google is to remove the line from `robots.txt`.

**Comment by rusefillc** [comment]
LIne 11 of https://github.com/robots.txt does not mention wiki anymore

Actually wiki is no longer mentioned at https://github.com/robots.txt?

**Reply by nelsonjchen** [reply]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

**FINALLY**! I will decommission https://github-wiki-see.page and just make it a pure-redirector now.

EDIT: SPOILERS: too early lmao. it's still blocked: https://github.com/github/feedback/discussions/4992#discussioncomment-1448177

**Reply by nelsonjchen** [reply]
GitHub Wiki SEE is now decommissioned and the tool will now redirect any old URLs pointing to https://github-wiki-see.page/ and whatnot to GitHub properly. 

@kevinjwalters , @gwlucastrig, @ann0see and others from from the old isaacs thread, the nightmare of getting our GitHub wiki content indexed on Google is now over.

**Reply by nelsonjchen** [reply]
![image](https://user-images.githubusercontent.com/5363/136305218-c5fb2611-66b0-4ba1-85ea-ebdad1983d00.png)

**Reply by nelsonjchen** [reply]
Sent in a PR to remove the statement in the docs: https://github.com/github/docs/pull/10836 .

**Reply by ann0see** [reply]
Great! Thank you GH and @nelsonjchen for the service you provided. We now moved to GH pages, but for sure this is a step in the right direction

**Reply by nelsonjchen** [reply]
No kidding! It's nice to see they opted to change the policy after nearly a decade.

**Reply by nelsonjchen** [reply]
Oh, um, nevermind. 😖

**Reply by nelsonjchen** [reply]
The nightmare continues.

**Comment by dipree** [comment, marked as answer]
Hi everyone :wave: we have intentionally excluded `Disallow: /*/wiki*` from the `robots.txt`. However, we have also introduced an `x-robots-tag: none` in the http response header of Wiki pages. As a result, Wikis are still not visible to search engine crawlers. Why have we done this? Abusive behavior in Wikis had a negative impact on our search engine ranking and therefore we had to exclude Wikis from getting crawled to mitigate the effects. We are now investigating options how we can open the gates again so that everyone can benefit from the great information documented in Wikis. At this point, we have not decided on whether or when we will allow Wikis to be crawled again, but we are actively reviewing this and will keep you posted with any updates.

**Reply by holta** [reply]
Can someone remind me what [Facebook Data Scientist Frances Haugen](https://en.wikipedia.org/wiki/Frances_Haugen) told the US Senate 3 days ago &mdash; about Big Tech putting profits above users ?

Thank you to all Microsoft employees who are likewise willing to put their stock options aside.
To step forward with a conscience as Frances did.
_To Do The Right Thing._

**Reply by dipree** [reply]
We appreciate that this group here is wanting your Wikis to appear in search results, and we want them to as well. It is unfortunate that the spam tactics have such a high impact. The SEO team at GitHub wants to allow search engines to crawl and rank the Wiki pages, however, we must be careful so that we do not end up with a Manual Action from any search engine due to spamminess. Please bear with us, and be patient as we continue to test for the best solution.

**Reply by mck1117** [reply]
So why not add a setting per repo to allow crawling?

**Reply by mahdyar** [reply]
@dipree Maybe enabling it for the repositories with more than n stars be a good solution?

**Reply by kevinjwalters** [reply]
At the risk of re-stating the obvious from my other ticket:

> It's still not clear to me why you wouldn't allow the wiki areas which are _not publicly editable_ to be available via Google Search and the like?

**Reply by nelsonjchen** [reply]
@kevinjwalters doing it on a per-page basis with headers instead of globally with `robots.txt` might be a start to that. Of course, I think some of the community would still like publicly editable wikis to be available to Google Search somehow though.

**Reply by SEOGoddess** [reply]
Thank you all for the feedback and ideas. We'll definitely keep this all into consideration as we work through this. To reiterate, the end goal here is to open up Wiki's to be crawled by search engines without risk of getting pegged for spam from those who abuse it.

**Reply by Caellian** [reply]
Here are some ideas:
- Non-shady crawlers have their names in user agent string.
- All projects with 1k+ stars get automatically whitelisted for crawling.
- Others get filtered based on number of links and whether they point to pages on a whitelist (e.g. Stack Overflow or any GitHub domain/subdomain)
  - It's safe to allow crawling of wikis without any links.
  - It's safe to allow crawling of wikis containing only links starting with a GitHub owned domain, if those links aren't blocked as well.
  - It's safe to scrub wiki content of links when rendering for crawler user agent.

EDIT: Removed stupidity & added "here's ideas" remark so it's clear I'm not associated with GitHub/MS.

**Reply by nelsonjchen** [reply]
Hmm, I've noticed that some URLs/sites like https://github.com/imchillin/Anamnesis/wiki/What-is-Anamnesis%3F are now indexed. It shows up on Google and the Network tab in chrome reports it doesn't have `x-robots-tag: none` on it. Is it part of some test? What's the criteria? It's certainly not by stars only as I'm comparing the ~300 stars to https://github.com/commaai/openpilot/wiki with 20k stars which still has `x-robots-tag: none`.

**Reply by nelsonjchen** [reply]
There are actually quite a few GitHub Wiki URLs that are now indexed: https://www.google.com/search?q=site:github.com+wiki

**Reply by nelsonjchen** [reply]
Oh, it looks like the criteria is currently if the wiki is publically editable or not. If it isn't, no `x-robots-tag: none`. That would be related to @kevinjwalters's remark. I'll try a test later.

**Reply by nelsonjchen** [reply]
It's not just that either, hmm. It's "Restrict Editing to Collaborators Only" (publically *un*-editable) and other unknown criteria(s) (stars?) at the moment.

**Reply by nelsonjchen** [reply]
The current criteria does appear to be "Restrict Editing to Collaborators Only" and some sort of user/organization aggregated star total. That said, there are odd stragglers left out here and there that don't meet the aggregated star requirement so I don't know what's going on there. There are also organizations such as Azure which meet the "requirements" but aren't permitted indexing as well. I saw all the Microsoft but realized the separate Azure org never made an appearance on top of the list.

About 1.2k wikis meet this rough criteria out of ~400k. 

https://docs.google.com/spreadsheets/d/1_W9s8IhnSoypXjTKYSRBK5CNPTQJfXRl7V_1mY0Xq28/edit#gid=905984022

Awesome! It's a start but hopefully GitHub will expand the list.

**Reply by nelsonjchen** [reply]
There are now 3.3k directly indexable wikis. The last spreadsheet seemed to favor some handpicked organizations but the new criteria seems to be stars. All seem to be ~700 stars or above.

All the directly indexable wikis are still un-editable except by collaborators from the small sample I looked at. 

Also whatever this new criteria is, it also dropped many wikis that were originally un-noindexed that fell in that weird handpicked-ish criteria in January. The [diff is ... diffy](https://github.com/nelsonjchen/github-wiki-see-rs/commit/5616d63a97193f9eb92ed22ec9e31a03b4cc085f).

https://docs.google.com/spreadsheets/d/1y5V1xVtL-Z4w4yt9mVQuwGCP3s8TgNAS756IKStRUqs/edit#gid=611938779

Awesome! It's cool to see the experimenting going on. Let's hope GitHub expands the list more.

**Reply by nelsonjchen** [reply]
The lowest number of stars of an indexable wiki is 793.

There are about 4753+ un-directly indexable wikis on repos with star counts equal to or above 793.

https://docs.google.com/spreadsheets/d/1HqjQNsJr2tTQ5XCXR_lpoYmmm9qrSrenvXF8kig1dNw/edit#gid=1520335241

**Reply by gwlucastrig** [reply]
I can see a counter argument for the star count criteria.  It seems to me that the projects that most need to be search-enabled are those that aren't already well known. A project with a high star count probably already has links on lots of web pages.  Now it is true, absolutely true, that the high-star projects ought to be indexed. But the place where being search-enabled can provide the most benefit is for those projects that serve a niche application and haven't yet been discovered by a lot of developers.

**Reply by Anouomgharbivsseagjdzzvz** [reply]
Qweqwe4844
content://media/external/file/3567

**Reply by 1989shack** [reply]
Okay, I hope in the nearest future we will begin to see some changes...

**Reply by nelsonjchen** [reply]
Update: ~500 star wikis that are non-publically-editable appear to be indexable now: https://github.com/IgniteUI/igniteui-angular/wiki 

No `x-robots-tag: none` present.

`x-robots-tag: none` is still present on publically-editable or wikis on repos with less than 500 stars.

**Reply by Novustat-Statistik-Beratung** [reply]
> but we are actively reviewing this and will keep you posted with any updates

Please announce changes before. We are currently adding a statistics wiki to our git profile which is also found on our website. If this would become crawlable for search engines, it would cause some major SEO issues (duplicate content) for us. However, allowing users to add html tags like rel="canonical" to wiki pages would solves this issue.

**Comment by jumpjack** [comment]
Wikis are not just the problem: google simply ignores ANY page in github, even readme.md! I can never find any suitable result in google coming from github. The only way for searching in github is using github search engine. But non-developers do not know github, so why should they search  there?
Try searching for "lunar lander game": there are some dozens of variants on github, no results in google!

**Reply by nelsonjchen** [reply]
Maybe you could just add `github` to your query if you're looking for GitHub variants. GitHub variants don't rank high because lunar lander game is something that have existed for decades with a rich history, ready to play variants (not necessarily GitHub/OSS) are what most people are looking for, and there's only so much space for results for Google to return in the first few pages.

**Reply by jumpjack** [reply]
it's just an example; github never comes up in generic google searches, although it contains thousands of programs, not just sources.

**Reply by dimpase** [reply]
perhaps things changed meanwhile, but https://www.google.com/search?q=%22lunar+lander+game%22+%22github%22+%22wiki%22
brings up GitHub wikis.

**Reply by nelsonjchen** [reply]
@dimpase https://github.com/orgs/community/discussions/4992#discussioncomment-3563126

**Comment by JackTrapper** [comment]
Fortunately there are some search engines that ignore the robot exclusion rule. They will in turn allow DuckDuckGo, Bing, and other bots to crawl their *cached* version of GitHub wiki pages - allowing us to find stuff on GitHub.

Its nice that someone cares more about finding content on the Internet

Now that the disallowing of robots has been rendered moot: lets do the right thing and remove any `robots.txt` or `x-robots-tag` headers and let information be available to as many people as possible - rather than caring about *"our search engine rankings"*.

**Comment by otykier** [comment]
Not sure what the conclusion of this discussion is, but we're facing this problem:

Around a year ago, I renamed my org from github.com/Otykier to github.com/TabularEditor. Now it seems, that none of the repo *issues* pages are being crawled. For example, I can take a literal quote from [this issue](https://github.com/TabularEditor/TabularEditor3/issues/187) (which was created more than a year ago) and get [no hits on Google](https://www.google.com/search?q=%22just+borrow+the+idea+from+SQL+Server+and+put+a+small+black+key%22). This goes for all issues - none of them seem to be crawled, making it very hard for users of our software to find solutions to problems they encounter. We can still get to the main page by searching for ["github tabular editor 3"](https://www.google.com/search?q=github+tabular+editor+3), for example, but no search results from the *issues* pages.

Is there some org or repo setting we're missing? I am 90% sure that it used to work before we renamed the org - could that have something to do with it? Thanks!

**Comment by DF4IAH** [comment]
I am thinking to move away from GitHub with this robots.txt policy  :-(

**Comment by jumpjack** [comment]
alternatives?

**Reply by DF4IAH** [reply]
I switched over to SourceForge.

**Reply by dimpase** [reply]
SF is not an alternative - they do not offer anything like GitHub actions, their GUI to work with code online is not existing, AFAIK.
The only real alternative is GitLab.

**Reply by ann0see** [reply]
We moved away from wikis to a website (hosted by GitHub Pages). I'd also consider GitLab an alternative.

**Comment by dimpase** [comment]
I am sure GitHub can easily arrange with at least one of the big search
engine to be well-crawled and indexed.

And this will give said engine a considerable boost in popularity.

**Reply by DF4IAH** [reply]
GitHub is now Microsoft owned. Don't know if they want their Azure in competition to GitHub. That is the reason that I move over to an alternative.

**Reply by dimpase** [reply]
azure does not do what github does, there is no competition; github uses azure to host everything they need, bringing extra business to azure.

**Comment by maxandersen** [comment]
We are seeing github wiki pages now being indexed - has this changed permanently or is it temporarily?

**Reply by nelsonjchen** [reply]
[Observationally](https://github-wiki-see.page):

* Above 500 stars
* "Editing Restricted to Repository Collaborators" turned on

**Comment by MaartenHilferink** [comment]
I asked OpenAI: https://chat.openai.com/c/308cf4b9-2759-41ca-a243-740279bbe1a7

**Reply by nelsonjchen** [reply]
You have to press "Share Conversation". Copy/pasting the URL won't work.

**Comment by ocornut** [comment]
I'm getting fairly desperate about various GitHub behaviors, but this is cherry on cake :(
It took me a long time to realize that energy spent on investing Wiki being a point documentation was hindered by the fact none of the info appears on e.g. Google, if it wasn't for a third-party creating a mirror of Wikis for that exact purpose. I have countless users searching for documentation bits that exists but aren't indexed 🤷 

My repo has 54K stars and in spite of a comment above suggesting that 500+ stars repos wikis would be indexed it doesn't seem like it is. *EDIT* I WOULD NEED TO MAKE MY WIKI UNEDITABLE BY USERS TO DO SO ?!!!

**Reply by nelsonjchen** [reply]
It must also not permit non-collaborators to edit in addition to the star requirement. I've noticed this criteria while developing that mirror thing. You can find a bit more info on the front page of that mirroring service https://github-wiki-see.page as discussions are always a bit fragmented. 


Does your repo also meet that criteria?

**Reply by ocornut** [reply]
It doesn’t, sorry i amended my post afterwards. It seems the betray the desirable purpose of my wiki to restrict edits. 

Thank you for your work on this Nelson .

**Reply by ocornut** [reply]
I need to clarify this didn't really help as well as I'd hoped, in spite of adding links here and there, I don't often stumble on the right mirrored info from a search engine.

**Comment by leandromsft** [comment]
I have a wiki in my repository containing documentation and coding standards for my repository.
I would like Copilot chat (Web Interface) to index these files. It's possible?

**Comment by ocornut** [comment]
(reposting some details not as a threaded reply)

@dipree: This is really harmful and disappointing, actively discouraging people investing in Wiki. I can't comprehend why it hasn't been solved. I understand that abuses are a problem, but a single "Allow only to collaborators" options is very limiting, other mitigating options should be possible (see below).

I have 56K+ stars, a growing Wiki https://github.com/ocornut/imgui/wiki and I would like other people than me to work on it, but right now its contents is not indexed in spite of best effort to even link to https://github-wiki-see.page etc. meaning **people are having difficulty finding the information they need, and keep asing us the same things, lowering the quality of their software experience and hogging support/dev resources.** :(

Can't github implement some alternative options? Suggestions in order of simplicity:
- a white list of users (who are not necessarily code collaborators)
- a system for users to request approval
- a system to let collaborator approve on first edit (~queued patches: I understand it is limiting in the sense that it means non-approved users wouldn't be able to make successive edits to queue, perhaps the UI can warn them beforehand).

If even the first option was added, I would cave in and lock the wiki tomorrow and add likely contributors + infos to request.

I suspect that part of your underlying thought might be that by requiring developers to grant full access to a repository to allow people edit a wiki, it is ensuring we don't add too many people? Can't exceptions be made to well behaving projects?

**Reply by nelsonjchen** [reply]
Unfortunately, the system, for better or worse, isn't really ranking that high on search engines nowadays. 

Even I find it pretty darn hard to use Google to look up openpilot wiki information which was the original impetus.

I'm thinking of maybe producing some sort of self mirroring toolkit for projects. Like a recipe for what the shellcheck project did. The existing system will still be there for general wikis though.

**Reply by 0xDE57** [reply]
Didn't realize that wiki's are not indexed. I just stumbled upon my own wiki in a search result hosted via GitHub Wiki Search Engine Enablement (GHWSEE).  (thanks for this useful resource!)

I had started putting things into wiki specifically so it could be indexed for others, but it appears not to be. That's unfortunate.

**Reply by liux674** [reply]
Yesterday I created a column on [Wiki](https://www.mediawiki.org/wiki/Talk:Communication#Traditional_vs_Laser_Cutting_for_Copper_Pipes_in_Hot_Weather_%E2%80%93_Any_UK_Experts?) about the application of [laser cutting](https://yihailasers.com/) in air conditioners in Europe.
